test/e2e/scenario_upgrade.go

```diff
@@ -52,6 +52,10 @@ func (scenario scenarioUpgrade) Title() string { return titleize(scenario.Name)
 
 func (scenario *scenarioUpgrade) SetInfra(infra Infra) {
 	scenario.infra = infra
+
+	// We want to use previous stable Terraform configs in upgrade tests
+	// as output format might get changed between different minor releases
+	scenario.infra.terraform = scenario.infra.terraformPrevious
```

Comment:
> People would init the cluster using "current at that time" terraform (i.e. stable), and then at the time of upgrade they allegedly will move to newer terraform config (tip).

I think the answer is _it depends_. We don't know for sure. I think it's safe to assume that users will only change `output.tf`, but leave everything else as-is. One example is that users don't want to change AMI or introduce some bigger architecture changes because that _might_ break their cluster. This was at least our assumption when introducing more significant changes to Terraform configs.

> This will cause to always use "stable" terraform, for both install and upgrade.

We do not run Terraform on upgrade time -- we run it only once when creating the infra and before provisioning the cluster.

I see your point here, but I don't think we should care much about this at the upgrade time. And I think we should try not to overcomplicate upgrade tests.